### PR TITLE
Add signature to display of functions in the shell

### LINKED ISSUE #9310: Define a better displayFormatter of function and methods
I think it should be possible to define a default formatter for functions that would be better than the defautl one. 

The following is not really useful:

```
In [1]: from inspect import signature
In [2]: signature
Out[2]: <function inspect.signature>
```

Something that would display less than `signature?` or `signature??`, but still a bit better than plain function/methods REPR. 

```
In [1]: from inspect import signature
In [2]: signature
Out[2]: <function inspect.signature(obj, *, follow_wrapped=True)>
```

Seem like a nice standalone project that could be useful. 

Thoughts ?

### PR DESCRIPTION
**Pull request by Madhu94**
Closes #9310

### CODE REVIEW

**Comment by takluyver** [comment]
This seems like a reasonable addition to me, but you'll need to update the test that's currently failing.

**Comment by Madhu94** [comment]
Sure, I'm really sorry, next time I'll make sure to check the tests.

**Comment by takluyver** [comment]
No problem, that's what CI is for ;-)

**Comment by Madhu94** [comment]
@takluyver Do you have any suggestions for dealing with outputs that vary across Python distributions? 

The output of ``inspect.signature(zlib.compress)`` is different in Python versions 3.4/3.5 and 3.6. That's why the updated test case passes in only 3.5 (which is what I use)

For Python 3.4 and Python3.5
```  
>>> str(inspect.signature(zlib.compress))    
'(bytes, level=-1, /)'
```

Python 3.6
```  
>>> str(inspect.signature(zlib.compress))
'(data, /, level=-1)'
```

Likewise for Python versions < 3.5, ``inspect.signature(id)`` will throw a ``ValueError``.

**Comment by takluyver** [comment]
A function defined in Python is more likely to have a reliable signature than a compiled one, I think. Try e.g. `posixpath.dirname`.

**Comment by Madhu94** [comment]
By "compiled one", you mean a ``builtin`` writtin in C ?  Got it, thanks.

**Comment by takluyver** [comment]
Yup, exactly.

**Comment by takluyver** [comment]
Ping! Do you think you'll get a chance to fix the test soon?

**Comment by Madhu94** [comment]
I had issues with my laptop and was offline for a while, will get to it
ASAP. It is a ridiculously small PR to be hanging around for so long.

**Comment by takluyver** [comment]
pickle has compiled functions, so you're still running into the same problem. I'd suggest `posixpath.dirname` as the example function - that's pure Python, it only takes one argument, and it's so basic that it's unlikely to change in the future.

**Comment by takluyver** [comment]
Or I guess you could define a function inside the test.

**Comment by Madhu94** [comment]
Yep. This time I will test on 3.3, 3.4, 3.5, 3.6 before pushing...:-/

**Comment by Madhu94** [comment]
@takluyver Are these tests fine ? I went with posixpath.dirname and my own function.

File: IPython/core/tests/test_oinspect.py [hunk 1]
```diff
@@ -38,14 +40,14 @@
 # defined, if any code is inserted above, the following line will need to be
 # updated.  Do NOT insert any whitespace between the next line and the function
 # definition below.
-THIS_LINE_NUMBER = 41  # Put here the actual number of this line
+THIS_LINE_NUMBER = 43  # Put here the actual number of this line
```
Comment (takluyver) [inline IPython/core/tests/test_oinspect.py]: I think this and the bits above are the result of a mistaken merge with another PR that got merged - can you set them back so they're unchanged in this PR?

**Review by takluyver — COMMENTED** [review]
Thanks, this is looking good for fixing the existing tests, but we should still add a test for the new functionality in IPython.lib.tests.test_pretty.

**Comment by Madhu94** [comment]
I had a question about that - why is [this test](https://github.com/ipython/ipython/blob/master/IPython/lib/tests/test_pretty.py#L223) called ``test_**un**bound_method`` ?

**Comment by takluyver** [comment]
```python
class MyClass:
    def amethod(self):
        pass

MyClass.amethod  # unbound method

instance = MyClass()
# bound method - it's bound to the instance, which becomes self when it is called
instance.amethod
```

In Python 2, unbound methods had their own type. In Python 3, they are just considered functions.

**Comment by takluyver** [comment]
Closing and reopening to re-run tests

(For some reason I don't have access to rebuild on Appveyor - @Carreau if you have access, can you see a way to add me?)

**Comment by Madhu94** [comment]
@takluyver It might have been better if I had gotten everything to work, and then asked you to merge - you did not have to check each commit. :)

**Comment by takluyver** [comment]
Sorry! I saw this was almost there, and didn't know if you would be available to finish it quickly.

**Comment by Madhu94** [comment]
I'll have a faster turnaround time on my next PR to ipython

**Comment by takluyver** [comment]
It hasn't been a problem, I was just a bit eager. Anyway, the tests are passing now, so I'm going to merge this. Thanks for your contribution!

**Pull request merged**